### ISSUE TITLE
[CLI] Parsing hyperparam config appears to be broken in 0.10.32, works in 0.10.29

### ISSUE THREAD

**Issue by nsfabina** [wandb/client#2317]
**Description**
Parsing of hyperparam configs for sweeps appears to be broken in 0.10.32. I reverted to 0.10.29 and it works again. Don't know in which bugfix version the error was introduced. Here are the warnings I get:

```wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways.
wandb: WARNING To avoid this, please fix the sweep config schema violations below:
wandb: WARNING   Violation 1. {'distribution': 'uniform', 'max': -18, 'min': -20} is not valid under any of the given schemas
wandb: WARNING   Violation 2. {'distribution': 'uniform', 'max': 0.01, 'min': 0} is not valid under any of the given schemas
wandb: WARNING   Violation 3. {'distribution': 'uniform', 'max': 0.05, 'min': 0} is not valid under any of the given schemas
wandb: WARNING   Violation 4. {'distribution': 'uniform', 'max': -18, 'min': -20} is not valid under any of the given schemas
wandb: WARNING   Violation 5. {'max': -18, 'min': -20, 'distribution': 'uniform'} is not valid under any of the given schemas
wandb: WARNING   Violation 6. {'distribution': 'uniform', 'max': 0.01, 'min': 0} is not valid under any of the given schemas
wandb: WARNING   Violation 7. {'distribution': 'uniform', 'max': -18, 'min': -20} is not valid under any of the given schemas
wandb: WARNING   Violation 8. {'distribution': 'uniform', 'max': -18, 'min': -20} is not valid under any of the given schemas
wandb: WARNING   Violation 9. {'distribution': 'uniform', 'max': 0.02, 'min': 0} is not valid under any of the given schemas
wandb: WARNING   Violation 10. {'max': -6, 'min': -8, 'distribution': 'uniform'} is not valid under any of the given schemas
wandb: WARNING   Violation 11. {'distribution': 'uniform', 'max': -18, 'min': -20} is not valid under any of the given schemas
```

Side-note 1:  I've seen config-parsing bugs fairly regularly in the wandb client. I've reported some in the past and I had actually pinned wandb at a much earlier version because config parsing was broken at some point. It might I decided to upgrade again recently and found this new parsing issue. I'm not sure if I'm missing a set of tests somewhere, but it seems like there are two locations where you all are testing wandb.sweep:  [here](https://github.com/wandb/client/blob/b181d10f518d356555925be21ea71f1b0c5bebdb/tests/wandb_sweep_test.py) and [here](https://github.com/wandb/client/blob/b181d10f518d356555925be21ea71f1b0c5bebdb/standalone_tests/sweep_check.py). It's probably worth writing a few tests that verify [the documented behaviors](https://docs.wandb.ai/guides/sweeps/configuration) are correct, as the current tests don't seem very thorough. 

Side-note 2: I'd also suggest raising when configs are not valid, rather than warning, and allowing users to pass in an option to ignore invalid config checks to bypass that error; that way, users aren't surprised when their deployments fail because of a missed warning and you still allow users to use working configs that don't meet your checks.

**Wandb features**
wandb.sweep

**How to reproduce**
Here's a config that didn't work in 0.10.32 but did work in 0.10.29.

```
{'entity': 'test', 'project': 'test', 'command': ['python', '-c', 'test'], 'method': 'bayes', 'metric': {'name': 'val_loss', 'goal': 'maximize'}, 'early_terminate': {'type': 'hyperband', 'min_iter': 8, 'eta': 2}, 'parameters': {'max_epochs': {'distribution': 'constant', 'value': 40}, 'optimizer': {'distribution': 'constant', 'value': 'adam'}, 'learning_rate_init_log': {'distribution': 'uniform', 'min': -8.0, 'max': -6.0}, 'learning_rate_factor': {'distribution': 'uniform', 'min': 0.75, 'max': 0.9}, 'learning_rate_init_mom': {'distribution': 'uniform', 'min': 0.9, 'max': 0.975}, 'loss_method': {'distribution': 'constant', 'value': 'rmse'}, 'augs_apply_rate': {'distribution': 'uniform', 'min': 0.05, 'max': 0.2}, 'augs_channel_dropout_rate': {'distribution': 'uniform', 'min': 0.0, 'max': 0.05}, 'augs_pixel_dropout_rate': {'distribution': 'uniform', 'min': 0.0, 'max': 0.02}, 'augs_pixel_noise_prob': {'distribution': 'uniform', 'min': 0.0, 'max': 0.01}, 'augs_pixel_noise_scale': {'distribution': 'uniform', 'min': 0.0, 'max': 0.01}, 'use_wgts': {'distribution': 'constant', 'value': 0.0}, 'block_structure': {'distribution': 'categorical', 'values': ['4,2,2', '2,2,2,1', '2,2,1']}, 'colorspace_filters_expand': {'distribution': 'categorical', 'values': [64, 96]}, 'colorspace_filters_contract': {'distribution': 'categorical', 'values': [32, 40]}, 'filters_initial': {'distribution': 'categorical', 'values': [96, 128]}, 'filters_growth': {'distribution': 'categorical', 'values': [16, 32]}, 'filters_final_expand': {'distribution': 'constant', 'value': 128}, 'filters_final_contract': {'distribution': 'constant', 'value': 64}, 'reg_act_l1_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_act_l2_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_bias_l1_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_bias_l2_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_kern_l1_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_kern_l2_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'use_maxmin_pool': {'distribution': 'constant', 'value': 'true'}}, 'name': 'test', 'description': 'test'}
```

**Environment**
- OS: OSX 11.4
- Python Version: 3.7.10

**Comment by tyomhak** [comment]
Hey @nsfabina , thank you for reporting! cc @raubitsj @dannygoldstein

**Comment by dannygoldstein** [comment]
Hi @nsfabina, what is happening here is that when you upsert a sweep schema, our backend validates it against a jsonschema to make sure it is properly formatted. Currently the JSONschema expects min/max limits for uniformly distributed parameters to be specified as floats instead of integers. This is to enable schema type inference. For example, on the backend, 

`{min: 0, max:10}` parses to a parameter distributed as `int_uniform`, whereas
`{min: 0.0, max: 10.0}` parses to a parameter distributed as `uniform` (floating point uniform). 

What I am going to do is in the case where you explicitly specify the uniform distribution

`{min: 0, max: 10, distribution: "uniform"}`

to allow min/max bounds to be specified as integers without throwing warnings.

Note: these warnings should not actually affect anything happening with your sweep, and it should still behave the way you expect. But I will put in something to disable those. Thanks for catching this edge case. 

The tests for the schema and config parsing live in the sweeps repo: https://github.com/wandb/sweeps

**Comment by dannygoldstein** [comment]
https://github.com/wandb/sweeps/blob/master/tests/test_schema.py
https://github.com/wandb/sweeps/blob/master/tests/test_validation.py

**Comment by nsfabina** [comment]
Hey @dannygoldstein 

I agree that your PR fixes the immediate issue here, but I think you've missed something about my issue that could potentially cause other problems. Your reply suggests that I've passed in integers and this is the root of the problem. This is incorrect. My config has floats in Python, e.g., `{'distribution': 'uniform', 'min': -20.0, 'max': -18.0}`. When the config is passed to `wandb.sweep`, the floats seem to be casted to integers at some point in the inner workings of the wandb pipeline. Please see my config below to confirm this.

The reason I bring this up is because, if this is the case, there may be other unintended consequences with how the wandb pipeline is handling the config values. That is, those floats are being converted to integers at some point and this PR handles the case where those values are checked for validity with the uniform distribution, but other issues may arise because of this incorrect casting.

Config here:

{'entity': 'test', 'project': 'test', 'command': ['python', '-c', 'test'], 'method': 'bayes', 'metric': {'name': 'val_loss', 'goal': 'maximize'}, 'early_terminate': {'type': 'hyperband', 'min_iter': 8, 'eta': 2}, 'parameters': {'max_epochs': {'distribution': 'constant', 'value': 40}, 'optimizer': {'distribution': 'constant', 'value': 'adam'}, 'learning_rate_init_log': {'distribution': 'uniform', 'min': -8.0, 'max': -6.0}, 'learning_rate_factor': {'distribution': 'uniform', 'min': 0.75, 'max': 0.9}, 'learning_rate_init_mom': {'distribution': 'uniform', 'min': 0.9, 'max': 0.975}, 'loss_method': {'distribution': 'constant', 'value': 'rmse'}, 'augs_apply_rate': {'distribution': 'uniform', 'min': 0.05, 'max': 0.2}, 'augs_channel_dropout_rate': {'distribution': 'uniform', 'min': 0.0, 'max': 0.05}, 'augs_pixel_dropout_rate': {'distribution': 'uniform', 'min': 0.0, 'max': 0.02}, 'augs_pixel_noise_prob': {'distribution': 'uniform', 'min': 0.0, 'max': 0.01}, 'augs_pixel_noise_scale': {'distribution': 'uniform', 'min': 0.0, 'max': 0.01}, 'use_wgts': {'distribution': 'constant', 'value': 0.0}, 'block_structure': {'distribution': 'categorical', 'values': ['4,2,2', '2,2,2,1', '2,2,1']}, 'colorspace_filters_expand': {'distribution': 'categorical', 'values': [64, 96]}, 'colorspace_filters_contract': {'distribution': 'categorical', 'values': [32, 40]}, 'filters_initial': {'distribution': 'categorical', 'values': [96, 128]}, 'filters_growth': {'distribution': 'categorical', 'values': [16, 32]}, 'filters_final_expand': {'distribution': 'constant', 'value': 128}, 'filters_final_contract': {'distribution': 'constant', 'value': 64}, 'reg_act_l1_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_act_l2_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_bias_l1_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_bias_l2_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_kern_l1_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'reg_kern_l2_log': {'distribution': 'uniform', 'min': -20.0, 'max': -18.0}, 'use_maxmin_pool': {'distribution': 'constant', 'value': 'true'}}, 'name': 'test', 'description': 'test'}

**Comment by dannygoldstein** [comment]
Hi @nsfabina thanks for following up !

I dug into this and found that it is actually a manifestation of this issue with a yaml library we use on our backend to parse sweep configs:

https://github.com/go-yaml/yaml/issues/430

I will put in a fix asap and let you know when it is in. 

Thanks!

**Comment by nsfabina** [comment]
Cool, I’m glad we could figure this out. Thanks for your patience and your
help. Have a good weekend.

**Comment by dannygoldstein** [comment]
This is now fixed on the backend
